src/java_bytecode/java_object_factory.cpp

```diff
@@ -593,21 +598,47 @@ codet initialize_nondet_string_struct(
         code_assumet(binary_relation_exprt(length_expr, ID_le, max_length)));
     }
 
+    // Data array needs to be dynamically allocated, to avoid reusing the
+    // same pointer in loops and function calls.
+    // char (*array_data_init)[INFINITY];
+    // array_data_init = allocate char[INFINITY]
+    symbolt &data_pointer_sym = get_fresh_aux_symbol(
+      pointer_type(
+        array_typet(java_char_type(), infinity_exprt(java_int_type()))),
+      "string_data_pointer",
+      "string_data_pointer",
+      loc,
+      ID_java,
+      symbol_table);
+    const auto data_pointer = data_pointer_sym.symbol_expr();
+    code.add(code_declt(data_pointer));
+
+    side_effect_exprt java_new_array(ID_allocate);
+    java_new_array.copy_to_operands(infinity_exprt(java_int_type()));
+    java_new_array.copy_to_operands(false_exprt());
+    java_new_array.type() = data_pointer.type();
```

Comment:
We should have/use a constructor instead of assigning fields in that way. (maybe add a todo).

Reply:
Or just pop off into another static `make_` function. Adding another constructor might not be necessary if this is domain specific.

Reply:
yes I constructor is maybe not the best way as there is no `allocate_exprt`, a `make_allocate_expr` could be useful if it is used in several places.